packages/starlight/user-components/Tabs.astro

```diff
@@ -9,8 +9,70 @@ interface Props {
 const { syncKey } = Astro.props;
 const panelHtml = await Astro.slots.render('default');
 const { html, panels } = processPanels(panelHtml);
+
+/**
+ * Synced tabs are persisted across page using `localStorage`. The script used to restore the
+ * active tab for a given sync key has a few requirements:
+ *
+ * - The script should only be included when at least one set of synced tabs is present on the page.
+ * - The script should be inlined to avoid a flash of invalid active tab.
+ * - The script should only be included once per page.
+ *
+ * To do so, we keep track of whether the script has been rendered using a variable stored using
+ * `Astro.locals` which will be reset for each new page. The value is not typed on purpose to avoid
+ * Starlight users to get autocomplete for it.
+ *
+ * The restore script defines a custom element `starlight-tabs-restore` that will be included in
+ * each set of synced tabs to restore the active tab based on the persisted value using the
+ * `connectCallback` lifecycle method. To ensure this callback can access all tabs and panels for
+ * the current set of tabs, the script should be rendered before the tabs themselves.
+ */
+// @ts-expect-error - See above
+const { didRenderSyncedTabsRestoreScript } = Astro.locals;
+const isSynced = syncKey !== undefined;
+const shouldRenderSyncedTabsRestoreScript = isSynced && didRenderSyncedTabsRestoreScript !== true;
+
+if (isSynced) {
+	// @ts-expect-error - See above
+	Astro.locals.didRenderSyncedTabsRestoreScript = true;
+}
 ---
 
+{/* Inlined to avoid a flash of invalid active tab. */}
+{shouldRenderSyncedTabsRestoreScript && <script is:inline>
+	(() => {
+		class StarlightTabsRestore extends HTMLElement {
+			connectedCallback() {
+				const starlightTabs = this.closest('starlight-tabs');
+				if (!(starlightTabs instanceof HTMLElement) || typeof localStorage === 'undefined') return;
+				const syncKey = starlightTabs.dataset.syncKey;
+				if (!syncKey) return;
+				let label = '';
+				try {
+					const storedSyncedTabs = localStorage.getItem('starlight-synced-tabs') ?? '{}';
+					const persistedSyncedTabs = JSON.parse(storedSyncedTabs);
+					if (typeof persistedSyncedTabs !== 'object') return;
+					label = persistedSyncedTabs[syncKey];
```

Comment:
Maybe I’m overthinking this but I’m thinking a bit about the storage model. This means there is a single JSON object in `localStorage` with state for all synced items and currently we re-read and re-parse it for every tab instance, right?

Almost feels like we should either:

1. Store this way but only read and parse once

2. Store each key separately (e.g. `localStorage.setItem('starlight-synced-tabs__' + syncKey, label)`) to make a single read simpler (less data per read/write, no JSON parsing)

I guess the second option would also simplify persistence because each `syncKey` only worries about its own state then — no need to handle reading and updating the full state object. The con is a “noisier” `localStorage` with potentially many keys. Would that be a concern do you think?

Reply:
Thanks for the early review, super appreciated.

I don't think you're overthinking this at all, and it's actually part of the improvements I mentioned in my previous message. My initial thought so far was closer to your second example but I may try for curiosity in a page with a lot of synced tabs if the first one can make a difference or not (don't think so but still want to test it now).